subsys/net/lib/mqtt/mqtt_transport_socket_tcp.c

```diff
@@ -29,26 +32,44 @@ int mqtt_client_tcp_connect(struct mqtt_client *client)
 	const struct sockaddr *broker = client->broker;
 	int ret;
 
-	client->transport.tcp.sock = socket(broker->sa_family, SOCK_STREAM,
-					    IPPROTO_TCP);
-	if (client->transport.tcp.sock < 0) {
-		return -errno;
-	}
+#if defined(CONFIG_SOCKS) && !defined(CONFIG_MQTT_LIB_TLS)
+	if (client->proxy == NULL)
+#endif
+	{
+		client->transport.tcp.sock = socket(broker->sa_family,
+						    SOCK_STREAM,
+						    IPPROTO_TCP);
+		if (client->transport.tcp.sock < 0) {
+			return -errno;
+		}
 
-	MQTT_TRC("Created socket %d", client->transport.tcp.sock);
+		MQTT_TRC("Created socket %d", client->transport.tcp.sock);
 
-	size_t peer_addr_size = sizeof(struct sockaddr_in6);
+		size_t peer_addr_size = sizeof(struct sockaddr_in6);
 
-	if (broker->sa_family == AF_INET) {
-		peer_addr_size = sizeof(struct sockaddr_in);
+		if (broker->sa_family == AF_INET) {
+			peer_addr_size = sizeof(struct sockaddr_in);
+		}
+
+		ret = connect(client->transport.tcp.sock, client->broker,
+			      peer_addr_size);
+		if (ret < 0) {
+			(void)close(client->transport.tcp.sock);
+			return -errno;
+		}
 	}
+#if defined(CONFIG_SOCKS) && !defined(CONFIG_MQTT_LIB_TLS)
```

Comment:
Could you use if (IS_ENABLED () && !IS_ENABLED()) ?

Reply:
@tbursztyka Thanks! Could you elaborate (or maybe point me to the right place in the doc) on when should `#if defined(CONFIG_*)` and `#if IS_ENABLED(CONFIG_*)` be used? I am trying to fix the PR based on the input from you and others, and this confuses me the most, as `if defined(...)` seems to be used throughout the current code (in headers, lib code, and the sample app code too).

Reply:
Well, we try to use IS_ENABLED as much as possible that's all. It does not mean we should never use #if defined(), there are places where you have to anyway.